editor/filesystem_dock.cpp

```diff
@@ -2881,6 +2984,8 @@ void FileSystemDock::_tree_rmb_select(const Vector2 &p_pos, MouseButton p_button
 	if (p_button != MouseButton::RIGHT) {
 		return;
 	}
+	tree->grab_focus();
```

Comment:
In my testing, right clicking tree items or file list items won't grab focus (left clicks do; right clicks don't). This feels unintentional, though, so it might be worth a separate PR to fix.

Behavior on the master branch:

https://github.com/godotengine/godot/assets/18225391/7314983a-cf91-44ec-9e69-18764ef3aeab

Reply:
Indeed, right clicks don't grab focus, see #75062
Though why is the focus needed here? Colors are changed for selected items, focus is irrelevant (?).

Reply:
I connected `id_pressed` for the folder colors submenu to its own method (`_folder_color_index_pressed`) instead of having `_file_option` handle it because I didn't want to fill up the FileMenu enum with options for all of the folder color items & end up having to hardcode the folder color submenu items (as opposed to letting it create them from the HashMap items, as it currently does). Although, that means it never receives the list of selected items like `_file_option` does, so I made it grab the list of selected items itself, and it needs which of the two menus are currently in focus for that.

Maybe I could write it so that it works similar to how the "New" submenu already works? I thought of doing that at first, but was hesitant to manually create all of the submenu options when it seemed more straightforward to do so by iterating the HashMap and checking the submenu item metadata in `_folder_color_index_pressed` instead.

Reply:
You could add only 2 enum values: `FOLDER_COLOR_DEFAULT` and `FOLDER_COLOR_MAX`, then instead of checking the `p_option` inside `switch` do `if (p_option >= FOLDER_COLOR_DEFAULT && p_option < FOLDER_COLOR_MAX)`. It's a bit hacky, but IMO acceptable. You could also add a comment that this is to avoid adding enum values for each color.

Reply:
Trying this out now - for the folder colors submenu to target the correct selections, I would need to connect `id_pressed` to either `_tree_rmb_option` or `_file_list_rmb_option` depending on which menu is in focus/was right-clicked anyways, wouldn't I? Because of that (as far as I can tell), the Create New submenu item doesn't even utilize the `p_selection` passed to `_file_option`, it uses the `current_path` variable of `FileSystemDock`, because it's always bound to `_tree_rmb_option` and wouldn't recieve the correct list of selections if this action was done on an item in the file list instead of the tree.

Perhaps I could use the existing system and look at `current_path` as the Create New submenu does, instead of checking focus and collecting selections as I'm doing now? It would mean only being able to set a folder color for one folder at a time, but I'm not sure if there are many scenarios where assigning many folders to the same color at a time would even be applicable, especially considering colors are inherited. Wouldn't have to mess with the enum, either.

Reply:
I can't believe a context menu is causing so many problems xd
I think restricting the option to single folder is fine. As you mentioned, folders inheriting colors make it a non-problem.

I think your `grab_focus()` solution could be fine too, but it changes behavior, so it needs opinion from more people (I don't think we grab focus on menus anywhere else in the editor). You could add a member `Control *context_menu_owner` set when calling `_file_and_folders_fill_popup()` and use it instead of checking focus.

Reply:
It might be beneficial to leave the `grab_focus()` solution in for now until #75062 is addressed, as it also fixes the rename issue (#78789). If it should be removed, I'll probably go ahead and use `current_path` instead to make it a little bit more straightforward.